state/indexer/indexer_service.go

```diff
@@ -75,25 +71,60 @@ func (is *Service) OnStart() error {
 				}
 			}
 
-			if err := is.blockIdxr.Index(eventDataHeader); err != nil {
-				is.Logger.Error("failed to index block", "height", height, "err", err)
-			} else {
-				is.Logger.Info("indexed block", "height", height)
+			if !IndexingEnabled(is.eventSinks) {
+				continue
 			}
 
-			if err = is.txIdxr.AddBatch(batch); err != nil {
-				is.Logger.Error("failed to index block txs", "height", height, "err", err)
-			} else {
-				is.Logger.Debug("indexed block txs", "height", height, "num_txs", eventDataHeader.NumTxs)
+			for _, sink := range is.eventSinks {
+				if err := sink.IndexBlockEvents(eventDataHeader); err != nil {
+					is.Logger.Error("failed to index block", "height", height, "err", err)
+				} else {
+					is.Logger.Info("indexed block", "height", height)
+				}
+
+				err := sink.IndexTxEvents(batch.Ops)
+				if err != nil {
+					is.Logger.Error("failed to index block txs", "height", height, "err", err)
+				}
 			}
 		}
 	}()
 	return nil
 }
 
-// OnStop implements service.Service by unsubscribing from all transactions.
+// OnStop implements service.Service by unsubscribing from all transactions and
+// close the eventsink.
 func (is *Service) OnStop() {
 	if is.eventBus.IsRunning() {
 		_ = is.eventBus.UnsubscribeAll(context.Background(), subscriber)
 	}
+
+	for _, sink := range is.eventSinks {
+		err := sink.Stop()
+		if err != nil {
+			is.Logger.Error("failed to close eventsink", "eventsink", sink.Type(), "err", err)
+		}
+	}
+}
+
+// KVSinkEnabled returns the given eventSinks is containing KVEventSink.
+func KVSinkEnabled(sinks []EventSink) bool {
+	for _, sink := range sinks {
+		if sink.Type() == KV {
+			return true
+		}
+	}
+
+	return false
+}
+
+// IndexingEnabled returns the given eventSinks is supporting the indexing services.
+func IndexingEnabled(sinks []EventSink) bool {
+	for _, sink := range sinks {
+		if sink.Type() == KV || sink.Type() == PSQL {
+			return true
+		}
+	}
```

Comment:
if sinks[0] == NULL ... ?

makes me think if we should associate an empty array with setting null instead of actually setting null.

Reply:
Do you mean we add a null eventsink when we see the tx-index config has empty settings?

Reply:
oh, I thought this is how it's done. If it's not, do we not set an event sink at all?

Reply:
right, I will update it. Thanks for the catch.